fix(preprod): Deduplicate sibling artifacts by app_id to prevent duplicate rows in status checks (EME-610)

runningcode opened the pull request:
## Summary

Fixes EME-610: Duplicate rows appearing in GitHub status checks after reprocessing/retriggering CI pipelines.

This PR addresses **only** the duplicate rows issue (not the duplicate check runs issue which is in PR #103521).

## Problem: Duplicate Rows After Reprocessing

### Issue
When CI is retriggered or artifacts are reprocessed, new artifacts are created with the same `app_id` as existing artifacts. The status check would then show ALL artifacts, resulting in duplicate rows.

**Example**:
- Initial upload: iOS-old, Android-old → **2 rows** in status check ✓
- Reprocess: iOS-new, Android-new created (4 total artifacts for same commit)
- Status check now shows: iOS-old, iOS-new, Android-old, Android-new → **4 rows** ❌

### Root Cause
`get_sibling_artifacts_for_commit()` returned ALL artifacts for a commit without deduplication:
```python
return PreprodArtifact.objects.filter(
    commit_comparison=self.commit_comparison,
    project__organization_id=self.project.organization_id,
).order_by("app_id")
```

### Solution
Deduplicate siblings by `app_id`:

1. Group all artifacts by `app_id`
2. For each `app_id` group:
   - If the **calling artifact** is in that group → use it
   - Otherwise → use the **earliest** (oldest) artifact for that app_id
3. Result: Max 1 artifact per `app_id`

**Example with fix**:
- When iOS-new triggers: Returns `[iOS-new, Android-old]` → **2 rows** ✓
- When Android-new triggers: Returns `[iOS-old, Android-new]` → **2 rows** ✓
- Never shows all 4 artifacts

---

## Testing

Added comprehensive test `test_sibling_deduplication_after_reprocessing`:
- Creates initial artifacts: iOS-old, Android-old
- Simulates reprocessing: iOS-new, Android-new (4 total)
- Verifies deduplication:
  - `ios_new.get_sibling_artifacts_for_commit()` → Returns [iOS-new, Android-old]
  - `android_new.get_sibling_artifacts_for_commit()` → Returns [iOS-old, Android-new]
  - Never returns all 4 artifacts

runningcode committed:
fix(preprod): Deduplicate sibling artifacts by app_id to prevent duplicate rows in status checks (EME-610)

This commit addresses Trevor's concern in EME-610: duplicate rows appearing in
status checks after reprocessing/retriggering CI pipelines.

Problem:
When CI is retriggered, new artifacts are created with the same app_id as
existing artifacts. The previous implementation returned ALL artifacts for a
commit, causing status checks to show duplicate rows (e.g., 4 rows after
reprocessing instead of 2).

Trevor's quote: "The status check itself isn't duplicated...The issue is the
duplicated content inside the status check, after re-running it went from 2x
builds -> 4x builds."

Solution:
Modified `get_sibling_artifacts_for_commit()` to deduplicate by app_id,
matching the behavior in the Emerge codebase (BaseBuildUtils.ts:findSiblingBuilds):
- Groups all artifacts by app_id
- For each app_id group:
  - If the calling artifact is in that group, use it
  - Otherwise, use the earliest (oldest) artifact for that app_id
- Result: Max 1 artifact per app_id, preventing duplicate rows

Example with fix:
- Initial upload: iOS-old, Android-old → 2 rows ✓
- Reprocess creates: iOS-new, Android-new (4 total artifacts)
- When iOS-new triggers: Returns [iOS-new, Android-old] → 2 rows ✓
- When Android-new triggers: Returns [iOS-old, Android-new] → 2 rows ✓

This ensures status checks consistently show one row per app, even after
multiple reprocessing runs.

Changes:
- Modified PreprodArtifact.get_sibling_artifacts_for_commit() to deduplicate
- Added comprehensive test test_sibling_deduplication_after_reprocessing

All 17 existing tests pass ✅

trevor-e commented on src/sentry/preprod/models.py:
```diff
@@ -144,21 +144,41 @@ def as_choices(cls) -> tuple[tuple[int, str], ...]:
+            else:
+                selected_ids.append(artifacts[0].id)
+
+        return PreprodArtifact.objects.filter(id__in=selected_ids).order_by("app_id")
```
Why is a second query needed afterwards, haven't we already fetched everything?

trevor-e commented on src/sentry/preprod/models.py:
```diff
@@ -144,21 +144,41 @@ def as_choices(cls) -> tuple[tuple[int, str], ...]:
-            QuerySet of PreprodArtifact objects, ordered by app_id for stable results
+            QuerySet of PreprodArtifact objects, deduplicated by app_id, ordered by app_id
         """
+        from collections import defaultdict
```
nit: AI loves to do this for some reason but keep all imports at the top unless necessary to break import loops

trevor-e reviewed (changes requested):
Think this looks good other than the platform change!

runningcode committed:
fix(preprod): Address PR review comments (EME-610)

- Deduplicate by (app_id, artifact_type) instead of just app_id to handle
  Android and iOS apps with the same app_id
- Move defaultdict import to top of file
- Add comment explaining why second query is needed
- Add test case for Android and iOS with same app_id
- Update existing test to set artifact_type fields
